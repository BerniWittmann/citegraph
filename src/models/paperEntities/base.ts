import { AuthorFields } from '@/models/paperEntities/author'
import { ConferenceFields } from '@/models/paperEntities/conference'
import { CountryFields } from '@/models/paperEntities/country'
import { InstitutionFields } from '@/models/paperEntities/institution'
import { PublicationFields } from '@/models/paperEntities/publication'
import { PublisherFields } from '@/models/paperEntities/publisher'
import { RecordFields } from '@/models/paperEntities/record'
import { constructQuery, PaperEntityQueryParameters } from '@/models/paperEntities/query'
import { constructMutation, PaperEntityMutationParameters } from '@/models/paperEntities/mutation'
import Transformer from '@/transformers/Transformer'

export interface PaperEntityBaseFields {
  id: string;
}

export default abstract class PaperEntity implements PaperEntityBaseFields {
  id: string;
  static readonly key: string;
  static readonly schemaName: string;
  static readonly queryName: string;
  static readonly displayedColumns: Array<PaperEntityTableColumn>;
  static readonly queryFields: string;
  static readonly mutationFields: Array<string> = [];
  static readonly transformer: typeof Transformer = Transformer;

  protected constructor ({ id }: PaperEntityBaseFields) {
    this.id = id
  }

  static getQuery (params: PaperEntityQueryParameters): string {
    return constructQuery(this.queryName, this.schemaName, params, this.queryFields)
  }

  static getMutation (params: PaperEntityMutationParameters, data: Partial<PaperEntityFields>): string {
    return constructMutation(this.queryName, this.schemaName, params, this.mutationFields, data)
  }
}

export interface PaperEntityTableColumn {
  text: string;
  value: string;
  align?: 'start' | 'center' | 'end'
  sortable?: boolean
  filterable?: boolean
  divider?: boolean
  class?: string | string[]
  width?: string | number
  filter?: (value: any, search: string, item: any) => boolean
  sort?: (a: any, b: any) => number,
  displayedByDefault?: boolean
}

export type PaperEntityFields = PaperEntityBaseFields & (
  AuthorFields &
  ConferenceFields &
  CountryFields &
  InstitutionFields &
  PublicationFields &
  PublisherFields &
  RecordFields
)
