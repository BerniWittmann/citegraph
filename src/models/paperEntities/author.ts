import PaperEntity, { PaperEntityBaseFields, PaperEntityTableColumn } from './base'
import AuthorTransformer from '@/transformers/AuthorTransformer'

export interface AuthorFields {
  firstName: string
  lastName: string
  countRecords: number
}

export default class Author extends PaperEntity implements AuthorFields {
  static key: string = 'author'
  static schemaName: string = 'author'
  static queryName: string = 'Authors'
  static queryFields: string = `id,
          first_name,
          last_name,
          number_citations`
  static mutationFields: Array<string> = ['first_name', 'last_name']
  static displayedColumns: Array<PaperEntityTableColumn> = [{
    text: 'project.explore.table.headers.author.first_name',
    value: 'firstName',
    sortable: true,
    filterable: true,
    displayedByDefault: true
  }, {
    text: 'project.explore.table.headers.author.last_name',
    value: 'lastName',
    sortable: true,
    filterable: true,
    displayedByDefault: true
  }, {
    text: 'project.explore.table.headers.author.count_records',
    value: 'countRecords',
    sortable: true,
    align: 'center'
  }]
  static transformer = AuthorTransformer
  firstName: string
  lastName: string
  countRecords: number

  constructor ({ id, firstName, lastName, countRecords }: PaperEntityBaseFields & AuthorFields) {
    super({ id })
    this.firstName = firstName
    this.lastName = lastName
    this.countRecords = countRecords
  }
}
