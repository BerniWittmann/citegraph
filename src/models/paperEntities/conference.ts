import PaperEntity, { PaperEntityBaseFields, PaperEntityTableColumn } from './base'

export interface ConferenceFields {
  name: string
  countRecords: number
}

export default class Conference extends PaperEntity implements ConferenceFields {
  static key: string = 'conference'
  static schemaName: string = 'conference'
  static queryName: string = 'Conferences'
  static queryFields: string = `name,
          number_citations`
  static displayedColumns: Array<PaperEntityTableColumn> = []
  name: string
  countRecords: number

  constructor ({ id, name, countRecords }: PaperEntityBaseFields & ConferenceFields) {
    super({ id })
    this.name = name
    this.countRecords = countRecords
  }
}
