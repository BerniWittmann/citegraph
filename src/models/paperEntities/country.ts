import PaperEntity, { PaperEntityBaseFields, PaperEntityTableColumn } from './base'
import CountryTransformer from '@/transformers/CountryTransformer'

export interface CountryFields {
  name: string
  countRecords: number
  flagUrl?: string
}

export default class Country extends PaperEntity implements CountryFields {
  static key: string = 'country'
  static schemaName: string = 'country'
  static queryName: string = 'Countries'
  static queryFields: string = `name,
          number_citations,
          flag_url`
  static mutationFields: Array<string> = ['name', 'flag_url']
  static displayedColumns: Array<PaperEntityTableColumn> = [{
    text: 'project.explore.table.headers.country.name',
    value: 'name',
    sortable: true,
    filterable: true,
    displayedByDefault: true
  }, {
    text: 'project.explore.table.headers.country.flag',
    value: 'flagUrl',
    sortable: false,
    filterable: false,
    align: 'center'
  }, {
    text: 'project.explore.table.headers.country.count_records',
    value: 'countRecords',
    sortable: true,
    align: 'center'
  }]
  static transformer = CountryTransformer
  name: string
  countRecords: number
  flagUrl?: string

  constructor ({ id, name, countRecords, flagUrl }: PaperEntityBaseFields & CountryFields) {
    super({ id })
    this.name = name
    this.countRecords = countRecords
    this.flagUrl = flagUrl
  }
}
