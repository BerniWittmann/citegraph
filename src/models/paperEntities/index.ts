import Author from './author'
import Conference from './conference'
import Country from './country'
import Institution from './institution'
import Publication from './publication'
import Publisher from './publisher'
import Record from './record'
import PaperEntity, { PaperEntityTableColumn, PaperEntityFields } from './base'

export const entities = [
  Record,
  Author,
  Institution,
  Country,
  Publication,
  Publisher,
  Conference
]

export const entityKeysMap = {
  [Record.key]: Record,
  [Author.key]: Author,
  [Institution.key]: Institution,
  [Country.key]: Country,
  [Publication.key]: Publication,
  [Publisher.key]: Publisher,
  [Conference.key]: Conference
}

export const entityKeys: Array<string> = entities.map((entity) => entity.key)

export type PaperEntityTableColumn = PaperEntityTableColumn

export type PaperEntity = PaperEntity

export type PaperEntityFields = PaperEntityFields

export default {
  Author,
  Conference,
  Country,
  Institution,
  Publication,
  Publisher,
  Record
}
