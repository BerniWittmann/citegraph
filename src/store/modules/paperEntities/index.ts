import { Module } from 'vuex'
import { PaperEntitiesState } from './types'
import { RootState } from '@/store/types'
import { state } from './state'
import { actions } from './actions'
import { getters } from './getters'
import { mutations } from './mutations'

const namespaced: boolean = true

export const paperEntities: Module<PaperEntitiesState, RootState> = {
  namespaced,
  state,
  getters,
  actions,
  mutations
}
