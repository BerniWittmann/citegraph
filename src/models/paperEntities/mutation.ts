import { PaperEntityFields } from '@/models/paperEntities/base'

export interface PaperEntityMutationParameters {
  id: string
  projectId: number
  entityType: string
}

export type PaperEntityMutationResponse = PaperEntityFields

function createMutationParams (params: PaperEntityMutationParameters): string {
  return `(id: ${params.id})`
}

function createData (fields: Array<string>, data: Partial<PaperEntityFields>): string {
  const updates = []
  for (let field of fields) {
    if (data.hasOwnProperty(field)) {
      // @ts-ignore
      updates.push(`${field}: "${data[field]}"`)
    }
  }
  return updates.join(', \n')
}

export function constructMutation (queryName: string, schemaName: string, params: PaperEntityMutationParameters, fields: Array<string>, data: Partial<PaperEntityFields>): string {
  return `{
      ${queryName}${createMutationParams(params)} {
        ${schemaName} {
          ${createData(fields, data)}
        }
      }
    }`
}
