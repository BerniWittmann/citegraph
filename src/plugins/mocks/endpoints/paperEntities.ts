import db from '../db'
import MockAdapter from 'axios-mock-adapter'
import { RecordFields } from '@/models/paperEntities/record'
import { setupMocks as setupGraphQLMocks } from './graphql-helpers/endpoint'
import { QueryConfiguration } from './graphql-helpers/queries'
import { AuthorFields } from '@/models/paperEntities/author'
import { CountryFields } from '@/models/paperEntities/country'

const records = db.getCollection('records')
const authors = db.getCollection('authors')
const countries = db.getCollection('countries')

const queryResponseConfiguration: Record<string, QueryConfiguration> = {
  'Records': {
    schemaName: 'record',
    queryName: 'Records',
    collection: records,
    filterFunction: recordsFilter
  },
  'Authors': {
    schemaName: 'author',
    queryName: 'Authors',
    collection: authors,
    filterFunction: authorsFilter
  },
  'Countries': {
    schemaName: 'country',
    queryName: 'Countries',
    collection: countries,
    filterFunction: countriesFilter
  }
}

function recordsFilter (obj: RecordFields, filter: string) {
  return obj.title.toLowerCase().includes(filter) ||
    obj.year.toString().includes(filter) ||
    obj.keywords.map(keyword => keyword.toLowerCase()).join('').includes(filter) ||
    obj.authors.map(author => (author.firstName + ' ' + author.lastName).toLowerCase()).join(';').includes(filter)
}

function authorsFilter (obj: AuthorFields, filter: string) {
  return obj.firstName.toLowerCase().includes(filter) ||
    obj.lastName.toLowerCase().includes(filter)
}

function countriesFilter (obj: CountryFields, filter: string) {
  return obj.name.toLowerCase().includes(filter)
}

export function setupMocks (mock: MockAdapter) {
  setupGraphQLMocks(mock, queryResponseConfiguration)
}
