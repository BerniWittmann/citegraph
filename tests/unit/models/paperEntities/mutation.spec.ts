import { constructMutation } from '@/models/paperEntities/mutation'

describe('models/paperEntities/mutation.ts', () => {
  describe('it has a function to create a mutation', () => {
    const params = {
      id: '42',
      projectId: 9,
      entityType: 'record'
    }
    it('creates a mutation', () => {
      const fields = ['id', 'name']
      const data = {
        id: '42',
        name: 'Name',
        keywords: ['too', 'much']
      }
      expect(constructMutation('Record', 'record', params, fields, data)).toMatchSnapshot()
    })

    it('handles invalid field declaration', () => {
      const fields = ['id', 'name', 'wrong']
      const data = {
        id: '42',
        name: 'Name',
        keywords: ['too', 'much']
      }
      expect(constructMutation('Record', 'record', params, fields, data)).toMatchSnapshot()
    })
  })
})
