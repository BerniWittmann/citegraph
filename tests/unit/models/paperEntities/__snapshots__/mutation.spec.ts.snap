// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models/paperEntities/mutation.ts it has a function to create a mutation creates a mutation 1`] = `
"{
      Record(id: 42) {
        record {
          id: \\"42\\", 
name: \\"Name\\"
        }
      }
    }"
`;

exports[`models/paperEntities/mutation.ts it has a function to create a mutation handles invalid field declaration 1`] = `
"{
      Record(id: 42) {
        record {
          id: \\"42\\", 
name: \\"Name\\"
        }
      }
    }"
`;
